core/evm.go

```diff
@@ -163,11 +135,25 @@ func newEVMBlockContext(header *types.Header, chain ChainContext, author *common
 		Header: &ethtypes.Header{
 			Number: new(big.Int).Set(header.Number),
 			Time:   header.Time,
-			Extra:  extra,
+			Extra:  header.Extra,
 		},
 	}
 }
 
+// NewEVMBlockContextWithPredicateResults creates a new context for use in the
+// EVM with an override for the predicate results. The miner uses this to pass
+// predicate results to the EVM when header.Extra is not fully formed yet.
+func NewEVMBlockContextWithPredicateResults(header *types.Header, chain ChainContext, author *common.Address, predicateResults []byte) vm.BlockContext {
+	blockCtx := NewEVMBlockContext(header, chain, author)
+	// Note this only sets the block context, which is the hand-off point for
+	// the EVM. The actual header is not modified.
+	blockCtx.Header.Extra = customheader.SetPredicateBytesInExtra(
+		blockCtx.Header.Extra,
```

Comment:
I think this might need to be slices.Clone(header.Extra) or something instead.. just to avoid altering header.Extra out of caution.